이미지 합성곱

-이미지는 2차원 -> 필터도 2차원 -> 현재 화소와 주위화소 배열을 각 요소 끼리 모두 곱하고 더함.  
-Zero padding : 바깥화소까지 처리해주기 위해 보통 0이나 1로 패딩함.  
-Pooling  
    -Max Pooling : 가장 큰 화소값 선택
    -Mean Pooling : 평균값 선택  
    -Min Pooling : 최솟값 선택
    -Overlapped Pooling : 

CNN 예시.
Convolution -> Pooling -> Convolution -> Pooling -> Fully Connected
Feature Extraction----------------------------------Classification

One-Hot-Encoding
각 카테고리 값을 이진 벡터로 표현.

#Conv2D는 2차원 이미지 데이터를 처리하기 위한 **합성곱 계층 (Convolutional Layer)**
-> 필터를 이용해 이미지 스캔
    •	filters: 출력 feature map의 수 (즉, 뉴런의 수)
	•	kernel_size: 필터의 크기 (예: (3,3)은 3x3 필터)
	•	activation: 활성화 함수 (relu, sigmoid 등)
	•	input_shape: (높이, 너비, 채널 수)


In [2]:
from tensorflow.keras import Sequential, layers
model = Sequential([
    layers.Input(shape=(28,28,1)), #인풋값 정의 3차원은 RGB값인데 이건 흓백.
    #뉴런 하나당 필터만큼의 가중치, 편향
    #-> 3x3x32 + 32 = 128개의 파라미터가 존재함.
    layers.Conv2D(32, (3,3), activation='relu'), #컨볼루션레이어, 32개
    layers.MaxPooling2D(), # pool_size=(2,2)가 기본값
    layers.Dropout(0.1),    #10퍼센트 뉴런 드롭아웃

    layers.Conv2D(64, 3, activation='relu'), #컨볼루션레이어, 64개
    layers.MaxPooling2D(), #Max 풀링
    layers.Dropout(0.25), #25퍼센트 드롭아웃

    layers.Flatten(), #1차원으로 바꾸고
    layers.Dense(128, activation='relu'),#Dense 레이어 128개

    layers.Dropout(0.5), #드롭아웃 50퍼센트
    
    layers.Dense(10, activation='softmax') #소프트맥스로 출력
    ])
model.summary()

2025-05-31 19:34:56.173506: I metal_plugin/src/device/metal_device.cc:1154] Metal device set to: Apple M2
2025-05-31 19:34:56.173543: I metal_plugin/src/device/metal_device.cc:296] systemMemory: 8.00 GB
2025-05-31 19:34:56.173556: I metal_plugin/src/device/metal_device.cc:313] maxCacheSize: 2.67 GB
I0000 00:00:1748687696.173742 1166798 pluggable_device_factory.cc:305] Could not identify NUMA node of platform GPU ID 0, defaulting to 0. Your kernel may not have been built with NUMA support.
I0000 00:00:1748687696.174254 1166798 pluggable_device_factory.cc:271] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 0 MB memory) -> physical PluggableDevice (device: 0, name: METAL, pci bus id: <undefined>)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 26, 26, 32)     │           320 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 13, 13, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 13, 13, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 11, 11, 64)     │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_1 (MaxPooling2D)  │ (None, 5, 5, 64)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 5, 5, 64)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 1600)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 128)            │       204,928 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_2 (Dropout)             │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 225,034 (879.04 KB)

 Trainable params: 225,034 (879.04 KB)

 Non-trainable params: 0 (0.00 B)

| 레이어 이름            | 출력 크기 (Output Shape) | 파라미터 수 (Param #)         | 설명 |
|------------------------|---------------------------|-------------------------------|------|
| Input                  | (28, 28, 1)               | 0                             | 입력 이미지 (28x28, 흑백) |
| Conv2D (1)             | (26, 26, 32)              | 320                           | (3×3×1)×32 + 32 = 320 |
| MaxPooling2D (1)       | (13, 13, 32)              | 0                             | 다운샘플링만 수행 |
| Dropout (0.1)          | (13, 13, 32)              | 0                             | 10% 뉴런 무작위 비활성화 |
| Conv2D (2)             | (11, 11, 64)              | 18,496                        | (3×3×32)×64 + 64 = 18,496 |
| MaxPooling2D (2)       | (5, 5, 64)                | 0                             | 다운샘플링만 수행 |
| Dropout (0.25)         | (5, 5, 64)                | 0                             | 25% 드롭아웃 |
| Flatten                | (1600,)                   | 0                             | 5×5×64 = 1600개로 펼침 |
| Dense (1)              | (128,)                    | 204,928                       | 1600×128 + 128 = 204,928 |
| Dropout (0.5)          | (128,)                    | 0                             | 50% 드롭아웃 |
| Dense (2 - Output)     | (10,)                     | 1,290                         | 128×10 + 10 = 1,290 |

**총 파라미터 수**: 225,034개

In [3]:
model.compile(loss='sparse_categorical_crossentropy' ,optimizer='adam', metrics=['accuracy'])

In [ ]:
from tensorflow.keras.datasets import mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()
X_train = X_train.reshape(-1, 28, 28, 1) / 255.0 #채널퍼스트방식.
X_test = X_test.reshape(-1, 28, 28, 1) / 255.0

In [5]:
model.fit(X_train, y_train, batch_size=128, epochs=10, verbose=0)
#배치 사이즈 ; 한번에 모델에 입력되는 데이터의 개수
#버보스 : 학습 진행 상황 출력

2025-05-31 19:36:29.974654: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:117] Plugin optimizer for device_type GPU is enabled.


In [6]:
loss, accuracy = model.evaluate(X_test, y_test, verbose=0)
print(f'Test loss: {loss}, Test accuracy: {accuracy}')

Test loss: 0.09701330214738846, Test accuracy: 0.9847000241279602


LeNet : 최초의 CNN
    convolution - subsampling(average pooling) - convolution - subsampling - full connection - Gaussian connection

In [ ]:
class LeNet:
    @staticmethod
    def build(input_shape=(32,32,1), activation='sigmoid'):
        model = Sequential()
        model.add(Input(shape=input_shape))